hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java

```diff
@@ -1997,23 +2015,31 @@ static String toStringHeader(ByteBuff buf) throws IOException {
       + onDiskDataSizeWithHeader;
   }
 
-  private static HFileBlockBuilder createBuilder(HFileBlock blk) {
+  /**
+   * Creates a new HFileBlockBuilder from the existing block and a new ByteBuff. The builder will be
+   * loaded with all of the original fields from blk, except now using the newBuff and setting
+   * isSharedMem based on the source of the passed in newBuff. An existing HFileBlock may have been
+   * an {@link ExclusiveMemHFileBlock}, but the new buffer might call for a
+   * {@link SharedMemHFileBlock}. Or vice versa.
+   * @param blk     the block to clone from
+   * @param newBuff the new buffer to use
+   */
+  private static HFileBlockBuilder createBuilder(HFileBlock blk, ByteBuff newBuff) {
```

Comment:
OK, so here is the trick.

In the old code, we will just reuse the ByteBuff when unpacking, and then call allocateBuffer to copy the content to a new buffer, and then decompress it.

And the problem is that, for the old implementation  of this method, we will always use the same isSharedMem, but actually, we may allocate off heap byte buff for the unpacked HFileBlock, while the packed HFileBlock is on heap.

Good.